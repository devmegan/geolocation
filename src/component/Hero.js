import { useState, useEffect } from "react";

import Description from "./Description";
import Error from "./Error";
import Geolocation from "./Location";
import AccuracyToggle from "./AccuracyToggle";

export default function Hero() {

    const [error, setError] = useState(false);
    const [getCoords, setGetCoords] = useState(false);
    const [coords, setCoords] = useState(false);
    const [isAccurate, setIsAccurate] = useState(false);
  
    const handleGeolocation = (geodata) => {
      setGetCoords(false);
      setCoords(false);
      
      if (geodata.error) {
        setError(geodata.error.message);
      } else {
        setError(false);
        setCoords(`${geodata.latitude}, ${geodata.longitude}`);
      }
    }
  
    const handleClick = () => {
      setGetCoords(true);
    }   
    const handleToggle = () => {
      setIsAccurate(!isAccurate);
    }

    return (
        <>
          <div className="hero min-h-screen bg-base-200">
            <div className="hero-content text-center">
              <div className="max-w-md">
                <Description coords={coords} />
                <button 
                  className="btn btn-primary btn-wide" 
                  onClick={handleClick}
                >
                  Get Location
                </button>
                {error && <Error error={error} setError={setError} />}
                <AccuracyToggle handleToggle={handleToggle} />
              </div>
            </div>
          </div>
          {getCoords && <Geolocation handleGeolocation={handleGeolocation} options={ { enableHighAccuracy: isAccurate } } />}
        </>
    )
}