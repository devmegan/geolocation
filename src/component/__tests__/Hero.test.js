import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Hero from '../Hero';

describe('hero', () => {
  it('renders w/o crashing', () => {
    expect.hasAssertions();

    const view = render(<Hero />);
    
    expect(view).toMatchSnapshot();
  });

  it('renders location component on click', () => {
    expect.hasAssertions();

    const mockGeolocation = {
        getCurrentPosition: jest.fn(),
        clearWatch: jest.fn(),
        watchPosition: jest.fn(),
    };

    global.navigator.geolocation = mockGeolocation;

    render(<Hero />);

    userEvent.click(screen.getByRole('button'));
    
    expect(mockGeolocation.getCurrentPosition).toHaveBeenCalledTimes(1);
  });

  it('renders error alert on geolocation error', () => {
    expect.hasAssertions();

    const mockGeolocation = {
        getCurrentPosition: (onChange, onError) =>
            onError(new Error('User denied Geolocation')),

        clearWatch: jest.fn(),
        watchPosition: jest.fn(),
    };

    global.navigator.geolocation = mockGeolocation;

    render(<Hero />);

    userEvent.click(screen.getByRole('button'));

    expect(screen.getByRole('alert')).toBeInTheDocument();
    expect(screen.getByRole('alert')).toHaveTextContent('User denied Geolocation');
  });

  it('removes error alert on successful geolocation response', () => {
    expect.hasAssertions();

    const mockGeolocationError = {
        getCurrentPosition: (onChange, onError) =>
            onError(new Error('User denied Geolocation')),

        clearWatch: jest.fn(),
        watchPosition: jest.fn(),
    };

    const mockGeolocationSuccess = {
        getCurrentPosition: jest.fn().mockImplementationOnce((success) =>
            Promise.resolve(
                success({
                    coords: {
                        latitude: 51.1,
                        longitude: 45.3,
                    },
                })
            )
        ),

        clearWatch: jest.fn(),
        watchPosition: jest.fn(),
    };

    global.navigator.geolocation = mockGeolocationError;

    render(<Hero />);
    
    const locationBtn = screen.getByRole('button');
    userEvent.click(locationBtn);

    expect(screen.getByRole('alert')).toBeInTheDocument();

    global.navigator.geolocation = mockGeolocationSuccess;

    userEvent.click(locationBtn);

    expect(screen.queryByRole('alert')).not.toBeInTheDocument();
  });
});
