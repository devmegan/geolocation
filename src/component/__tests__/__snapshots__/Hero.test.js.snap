// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hero renders w/o crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="hero min-h-screen bg-base-200"
      >
        <div
          class="hero-content text-center"
        >
          <div
            class="max-w-md"
          >
            <h1
              class="text-5xl font-bold"
            >
              useGeolocation
            </h1>
            <p
              class="py-6"
            >
              A React hook that grabs a user's coordinates.
            </p>
            <button
              class="btn btn-primary btn-wide"
            >
              Get Location
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="hero min-h-screen bg-base-200"
    >
      <div
        class="hero-content text-center"
      >
        <div
          class="max-w-md"
        >
          <h1
            class="text-5xl font-bold"
          >
            useGeolocation
          </h1>
          <p
            class="py-6"
          >
            A React hook that grabs a user's coordinates.
          </p>
          <button
            class="btn btn-primary btn-wide"
          >
            Get Location
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
